src/sidechain.cpp

```diff
@@ -0,0 +1,466 @@
+                CreateDBEntry(view, txundo, block_height, sidechain_record_id, proposal_s);
+                new_sidechains_activated.insert(proposal.idnum);
+            }
+            DeleteDBEntry(view, txundo, record_id);
+            record_id.n = DBIDX_SIDECHAIN_PROPOSAL;
+            DeleteDBEntry(view, txundo, record_id);
+        }
+
+        if (!new_sidechains_activated.empty()) {
+            Assume(sidechain_proposal_list.size() != sidechain_proposal_list_new.size());
+            // Assign CTIPs
+            for (size_t output_index = 0; output_index < tx.vout.size(); ++output_index) {
+                if (!tx.vout[output_index].scriptPubKey.IsDrivechain()) continue;
+
+                const uint8_t sidechain_id = tx.vout[output_index].scriptPubKey[DRIVECHAIN_SCRIPT_SIDECHAIN_ID_OFFSET];
+                if (new_sidechains_activated.find(sidechain_id) == new_sidechains_activated.end()) {
+                    // TODO: OP_DRIVECHAIN (or OP_NOP5) for a non-activating sidechain id; should this be invalid??
+                    continue;
+                }
+
+                CDataStream ctip_info(SER_NETWORK, PROTOCOL_VERSION);
+                ctip_info << sidechain_id;
+                CreateDBEntry(view, txundo, block_height, {tx.GetHash(), DBIDX_SIDECHAIN_CTIP_INFO}, ctip_info);
+            }
+            if (!new_sidechains_activated.empty()) {
+                return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-drivechain-activated-without-ctip");
+            }
+        }
+
+        if (sidechain_proposal_list.size() != sidechain_proposal_list_new.size()) {
+            Assume(!new_sidechains_activated.empty());
+            COutPoint record_id{ArithToUint256(arith_uint256{(uint64_t)completed_block_height}), DBIDX_SIDECHAIN_PROPOSAL_LIST};
+            DeleteDBEntry(view, txundo, record_id);
+
+            if (!(sidechain_proposal_list_new.empty() && withdraw_proposal_list.empty())) {
+                CDataStream proposal_list(SER_NETWORK, PROTOCOL_VERSION);
+                proposal_list << sidechain_proposal_list_new;
+                proposal_list << withdraw_proposal_list;
+                CreateDBEntry(view, txundo, block_height, record_id, proposal_list);
+            }
+        }
+    }
+
+    return true;
+}
+
+bool VerifyDrivechainSpend(const CTransaction& tx, const unsigned int sidechain_input_n, const std::vector<CTxOut>& spent_outputs, const CCoinsViewCache& view, TxValidationState& state) {
+    const CTxIn& sidechain_input = tx.vin[sidechain_input_n];
+    // TODO: Do we want to verify there's only one sidechain involved? BIP300 says yes, but why?
+
+    // Lookup sidechain number from CTIP and ensure this is in fact a CTIP to begin with
+    // FIXME: It might be a good idea to include the sidechain # in the tx itself somewhere?
+    uint8_t sidechain_id;
+    {
+        CDataStream ctip_info = GetDBEntry(view, {sidechain_input.prevout.hash, DBIDX_SIDECHAIN_CTIP_INFO});
```

Comment:
I think we need to update DBIDX_SIDECHAIN_CTIP_INFO with the new CTIP identified below either after deposit (L409) or withdrawal (L464). Otherwise we won't get the `ctip_info` here. 

Also, if we spend the current ctip to do a deposit, then that would mean a deposit would invalidate any current withdrawals in-flight because the withdrawal also has to spend the ctip?

Reply:
No, to the second question. The withdrawal has a version of APO (called the "blinded hash", with the CTIP input zeroed out).